test/fixtures/es-module-loaders/loader-with-custom-condition.mjs

```diff
@@ -0,0 +1,11 @@
+import {ok} from 'assert';
+
+export async function resolve(specifier, context, defaultResolve) {
+  ok(Array.isArray(context.conditions), 'loader receives conditions array');
+  ok(context.conditions.includes('node'), 'provided conditions include "node"');
```

Comment:
What do you think about calling this `envConditions` for "environment conditions" rather than "conditions"? Just thinking in terms of terminology if we want to try and narrow this down? Or `nodeEnv` or even just `env`?

Reply:
I'd avoid trying to state they are environmental, as I don't think that fits "require"

Reply:
`require` and `import` are very specific exceptions to the rule. And can still sort of be considered "static" down to the fact that they are loader-environment-specific.

Overall conditions should be considered environment static though...

Reply:
I disagree that they should be static and environmental

Reply:
I think environment is already very overloaded, especially if dev/prod ever becomes one of the conditions, making (node) env a subset of conditions. Conditions also matches "conditional exports" and I think it's too late to rename that one.

Reply:
What I'm arguing for here is two types of conditions:

1. "System conditions" - `"default"`, `"node"`, `"require"`, `"import"` (and later `"self"`) that cannot be changed or altered, since they apply based on their rules which are formally defined and to change those rules would be to break the contract.
2. "User environment conditions" - `process.env.NODE_ENV.split(',')` - that apply additionally to the system conditions, but not in any way that can interfere with them.

If we integrate a split like this, then I'd be happy to expose the `"user environment conditions"` to loaders through the context in this way, but I am against allowing loaders to break the contract of the system conditions in (1).

Reply:
> since they apply based on their rules which are formally defined and to change those rules would be to break the contract.

Can you clarify this? I'm not sure what contract is being talked about.

Reply:
Node.js should always match the `"node"` condition, since that is the definition of that condition. Packages can rely on this by using patterns like:

```js
{
  "exports": {
    "node": "./asdf.js",
    "default": "./DEFINITELY_NOT_NODEJS.js"
  }
}
```

and the contract in the above is that `"default"` will never ever be a Node.js main resolution.

I wouldn't want loaders to be able to just randomly change these set resolutions - but rather only to add new ones that don't have existing definitions for Node.js.

Reply:
If a loader is meaning to emulate a different environment than Node I would expect it to be able to change the defaults to not include the "node" condition. I'm still unclear on the contract here, but it might be because I do not see loaders as being constrained to only emulate Node.

Reply:
I think there's two different questions for me:

1. Should node allow loaders to mess with certain "reserved" conditions? We already restricted these by protecting `require`/`import` and not exposing `default`. We could expand this to the `node` conditions` (fail if the hook calls `defaultResolve` without it).
2. Should resolve hooks make assumptions about the conditions that are active? Should they hard-code that resolution includes `node` or `import` etc.?

Since I want to be able to use `resolve` hooks outside of node, I feel pretty strongly that (2) should *not* be the case. I would want to be able to reuse custom resolution hooks for non-node environments like bundlers. And the only way to allow that is if `node` is one of the arguments and not implicitly assumed by the hook.

Reply:
Right, and that's why I think this PR is an over-generalization of a more simple `process.env.NODE_ENV` conditional approach.

We are trying to solve for perfect abstraction and the usability of users adding conditions to their apps. Those are two separate use cases.